File: src/test/java/ModulusResultTests.java
import com.dambra.paul.moduluschecker.chain.ModulusResult;
import org.junit.Test;

import java.util.Optional;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

public class ModulusResultTests {
    @Test
    public void CanCopyFirstResult() {
        ModulusResult original = new ModulusResult(Optional.of(true), Optional.empty());

        ModulusResult copy = ModulusResult.copy(original);

        assertThat(copy.firstException.isPresent(), is(false));
        assertThat(copy.secondException.isPresent(), is(false));

        assertThat(copy.firstCheckResult.orElse(false), is(true));
        assertThat(copy.secondCheckResult.isPresent(), is(false));
    }

    @Test
    public void CanCopySecondResult() {
        ModulusResult original = new ModulusResult(Optional.of(true), Optional.of(false));

        ModulusResult copy = ModulusResult.copy(original);

        assertThat(copy.firstException.isPresent(), is(false));
        assertThat(copy.secondException.isPresent(), is(false));

        assertThat(copy.firstCheckResult.orElse(false), is(true));
        assertThat(copy.secondCheckResult.orElse(true), is(false));
    }

    @Test
    public void CanCopyFirstException() {
        ModulusResult original = new ModulusResult(Optional.of(true), Optional.of(false));
        final ModulusResult modulusResult = original.withFirstException(Optional.of(5));
        ModulusResult copy = ModulusResult.copy(modulusResult);

        assertThat(copy.firstException.get(), is(5));
        assertThat(copy.secondException.isPresent(), is(false));

        assertThat(copy.firstCheckResult.orElse(false), is(true));
        assertThat(copy.secondCheckResult.orElse(true), is(false));
    }

    @Test
    public void CanCopySecondException() {
        ModulusResult original = new ModulusResult(Optional.of(true), Optional.of(false));
        final ModulusResult modulusResult = original.withSecondException(Optional.of(10));
        ModulusResult copy = ModulusResult.copy(modulusResult);

        assertThat(copy.firstException.isPresent(), is(false));
        assertThat(copy.secondException.get(), is(10));

        assertThat(copy.firstCheckResult.orElse(false), is(true));
        assertThat(copy.secondCheckResult.orElse(true), is(false));
    }

    @Test
    public void CanCopyBothExceptions() {
        ModulusResult original = new ModulusResult(Optional.of(true), Optional.of(false));
        final ModulusResult x = original.withFirstException(Optional.of(3));
        final ModulusResult y = x.withSecondException(Optional.of(4));
        ModulusResult copy = ModulusResult.copy(y);

        assertThat(copy.firstException.get(), is(3));
        assertThat(copy.secondException.get(), is(4));

        assertThat(copy.firstCheckResult.orElse(false), is(true));
        assertThat(copy.secondCheckResult.orElse(true), is(false));
    }

    @Test
    public void CanCopyWithSecondResult() {
        ModulusResult original = new ModulusResult(Optional.of(true), Optional.empty());
        final ModulusResult x = original.withFirstException(Optional.of(6));
        final ModulusResult y = x.withSecondException(Optional.of(7));
        ModulusResult copy = ModulusResult.withSecondResult(Optional.ofNullable(y), true);

        assertThat(copy.firstException.get(), is(6));
        assertThat(copy.secondException.get(), is(7));

        assertThat(copy.firstCheckResult.orElse(false), is(true));
        assertThat(copy.secondCheckResult.orElse(false), is(true));
    }
}

File: src/test/java/ModulusResultTests.kt


import com.github.pauldambra.moduluschecker.chain.ModulusResult
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test

class ModulusResultTests {
    @Test
    fun canCopyFirstResult() {
        val original = ModulusResult(true, null)

        val copy = ModulusResult.copy(original)

        assertThat(copy?.firstException, `is`(null as Int?))
        assertThat(copy?.secondException, `is`(null as Int?))

        assertThat(copy?.firstCheckResult ?: false, `is`(true))
        assertThat(copy?.secondCheckResult, `is`(null as Boolean?))
    }

    @Test
    fun canCopySecondResult() {
        val original = ModulusResult(true, false)

        val copy = ModulusResult.copy(original)

        assertThat(copy?.firstException, `is`(null as Int?))
        assertThat(copy?.secondException, `is`(null as Int?))

        assertThat(copy?.firstCheckResult ?: false, `is`(true))
        assertThat(copy?.secondCheckResult ?: true, `is`(false))
    }

    @Test
    fun canCopyFirstException() {
        val original = ModulusResult(true, false)
        val modulusResult = original.withFirstException(5)
        val copy = ModulusResult.copy(modulusResult)

        assertThat(copy!!.firstException, `is`(5))
        assertThat(copy.secondException, `is`(null as Int?))

        assertThat(copy.firstCheckResult ?: false, `is`(true))
        assertThat(copy.secondCheckResult ?: true, `is`(false))
    }

    @Test
    fun canCopySecondException() {
        val original = ModulusResult(true, false)
        val modulusResult = original.withSecondException(10)
        val copy = ModulusResult.copy(modulusResult)

        assertThat(copy?.firstException, `is`(null as Int?))
        assertThat(copy!!.secondException, `is`(10))

        assertThat(copy.firstCheckResult ?: false, `is`(true))
        assertThat(copy.secondCheckResult ?: true, `is`(false))
    }

    @Test
    fun canCopyBothExceptions() {
        val original = ModulusResult(true, false)
        val x = original.withFirstException(3)
        val y = x.withSecondException(4)
        val copy = ModulusResult.copy(y)

        assertThat(copy!!.firstException, `is`(3))
        assertThat(copy.secondException, `is`(4))

        assertThat(copy.firstCheckResult ?: false, `is`(true))
        assertThat(copy.secondCheckResult ?: true, `is`(false))
    }

    @Test
    fun canCopyWithSecondResult() {
        val original = ModulusResult(true, null)
        val x = original.withFirstException(6)
        val y = x.withSecondException(7)
        val copy = ModulusResult.withSecondResult(y, true)

        assertThat(copy.firstException, `is`(6))
        assertThat(copy.secondException, `is`(7))

        assertThat(copy.firstCheckResult ?: false, `is`(true))
        assertThat(copy.secondCheckResult ?: false, `is`(true))
    }
}
